Implement CRD Establishment Utility Function

frankgreco opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://git.k8s.io/community/contributors/devel/pull-requests.md#the-pr-submit-process and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/devel/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/devel/pull-requests.md#write-release-notes-if-needed
4. If the PR is unfinished, see how to mark it: https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md#marking-unfinished-pull-requests
-->

**What this PR does / why we need it**:
Once a `CustomResourceDefinition` is created, it cannot be used until it has been *established*. This PR implements a utility function that will create a CRD(s) and wait for its/their establishment. 

**Which issue(s) this PR fixes**
Fixes [https://github.com/kubernetes/apiextensions-apiserver/issues/9](https://github.com/kubernetes/apiextensions-apiserver/issues/9)

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
NONE
```

Commit:
implement customresourcedefinition establishment utility function

implement customresourcedefinition establishment utility function

implement customresourcedefinition establishment utility function

implement customresourcedefinition establishment utility function

implement customresourcedefinition establishment utility function

implement customresourcedefinition establishment utility function

implement customresourcedefinition establishment utility function

implement customresourcedefinition establishment utility function

implement customresourcedefinition establishment utility function

implement customresourcedefinition establishment utility function

implement customresourcedefinition establishment utility function

initial commit of crd ensure utility function

formatting go files

fixing package name

removing coverage file

adding unit tests/refactor helpers

initial commit of crd ensure utility function

formatting go files

fixing package name

removing coverage file

adding unit tests/refactor helpers

fix copy/paste typo

fixing linting errors

adding proper boilerplate headers

fixing merge confists syntax

fixing merge conflicts syntax

formatting...

adding boilerplate header

implement aggregate errors equality helper

fix linting error

adding CRD establishment helper

moving helper files

removing redundant imports

adding CRD establishment helper

renaming files

adding CRD establishment helper

qualifying package variables

nikhita commented:
/ok-to-test

nikhita commented on staging/src/k8s.io/apiextensions-apiserver/pkg/utils/utils.go:
```diff
@@ -0,0 +1,84 @@
+
+var (
+	// retries is the maximum number of attempts that a CRD will be given to establish
+	retries int = 5
```
nit: the `int` can be dropped. just `retries = 5`.

nikhita commented on staging/src/k8s.io/apiextensions-apiserver/pkg/utils/utils.go:
```diff
@@ -0,0 +1,84 @@
+	// retries is the maximum number of attempts that a CRD will be given to establish
+	retries int = 5
+	// interval is the amount of time that will pass between retries
+	interval time.Duration = 500 * time.Millisecond
```
nit: time.Duration can be dropped.

nikhita commented on staging/src/k8s.io/apiextensions-apiserver/pkg/utils/utils.go:
```diff
@@ -0,0 +1,84 @@
+	// interval is the amount of time that will pass between retries
+	interval time.Duration = 500 * time.Millisecond
+	// factor is the scalar by which the prevous interval will be increased
+	factor float64 = 1.0
```
nit: float64 can be dropped.

nikhita commented on staging/src/k8s.io/apiextensions-apiserver/pkg/utils/utils.go:
```diff
@@ -0,0 +1,84 @@
+func EnsureCRDs(i apiextensionsv1beta1client.ApiextensionsV1beta1Interface, crds ...*apiextensionsv1beta1.CustomResourceDefinition) error {
+	funcs := make([]func() error, len(crds))
+	for index, crd := range crds {
+		crd := crd // We want to process every crd, not just the last one.
```
Maybe a deep copy here?

nikhita commented on staging/src/k8s.io/apiextensions-apiserver/pkg/utils/utils.go:
```diff
@@ -0,0 +1,84 @@
+		Duration: interval,
+	}, func() (bool, error) {
+		// Attempt to retrieve the CRD that was either already present or just created.
+		crd, err := i.CustomResourceDefinitions().Get(crd.ObjectMeta.Name, metav1.GetOptions{})
```
nit: `crd.GetName()`

nikhita commented on staging/src/k8s.io/apiextensions-apiserver/pkg/utils/utils.go:
```diff
@@ -0,0 +1,84 @@
+				// If we have reached this state, the CRD will never become
+				// established
+				if cond.Status == apiextensionsv1beta1.ConditionFalse {
+					return false, errors.New(cond.Reason)
```
This will just list the reason as the error. Maybe a more descriptive message would be better.

nikhita commented on staging/src/k8s.io/apiextensions-apiserver/pkg/utils/utils.go:
```diff
@@ -0,0 +1,84 @@
+// EnsureCRDs is a utility function that will ensure that a list of CustomResourceDefinitions
+// are not only created, but ready for use by a Kubernetes cluster. If a CustomResourceDefinition
+// cannot be established within a reasonable amount of retries, an error will be returned.
+func EnsureCRDs(i apiextensionsv1beta1client.ApiextensionsV1beta1Interface, crds ...*apiextensionsv1beta1.CustomResourceDefinition) error {
```
@frankgreco would be nice to have a test for this as well. :)

nikhita commented on staging/src/k8s.io/apiextensions-apiserver/pkg/utils/utils.go:
```diff
@@ -0,0 +1,84 @@
+package utils
```
There is a helpers.go already. It would probably be better to move those functions here into the utils package.

sttts replied:
See my comment above.

frankgreco replied:
helpers have been moved back.

Commit:
autogenerated BUILD files

autogenerated BUILD files

autogenerated BUILD files

autogenerated BUILD files

autogenerated BUILD files

autogenerated BUILD files

updating BUILD files with autogenerated ones

updating BUILD files

updating BUILD files

bug fix in BUILD

adding dep to BUILD

adding dep to BUILD

adding dep to BUILD

adding dep to BUILD

updating generated BUILD files

updating generated BUILD files

updating generated BUILD files

autogenerated BUILD files

autogenerated BUILD files

autogenerated BUILD files

autogenerated BUILD files

frankgreco commented:
/test pull-kubernetes-unit

sttts commented:
Please squash everything generated into one commit and everything into another one. Makes it much easier to review.

frankgreco commented:
@sttts commits are squashed as requested.

frankgreco commented:
@sttts @nikhita thoughts?

frankgreco commented:
@sttts @nikhita so I almost wonder if this `Ensure` function should live [here](https://github.com/kubernetes/apiextensions-apiserver/blob/master/pkg/client/clientset/clientset/typed/apiextensions/v1beta1/generated_expansion.go#L19).

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/utils/helpers.go:
```diff
@@ -14,17 +14,19 @@ See the License for the specific language governing permissions and
 */
 
-package apiextensions
+package utils
```
There are helpers for the api. Don't like to move them into some generic pkg/utils package. Why do we need that?

sttts replied:
Another reason for that: the helpers are about internal types. They shouldn't be used outside of the apiserver itself.

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/utils/utils.go:
```diff
@@ -0,0 +1,105 @@
+// EnsureCRDs is a utility function that will ensure that a list of CustomResourceDefinitions
+// are not only created, but ready for use by a Kubernetes cluster. If a CustomResourceDefinition
+// cannot be established within a reasonable amount of retries, an error will be returned.
+func EnsureCRDs(
```
the name does not say what it ensures. I would also expect to have a stopCh to cancel it.

sttts replied:
We traditionally don't move helpers into the clients. I tend to leave it separate.

Another point: why do we need the slice of CRDs in that func? IMO the func should be for one CRD and the user can do any concurrency magic around.

frankgreco replied:
There are some benefits with adding it to the expansion interface:
* Can be versioned with the rest of the API. By keeping it separate in a utils pkg, you can only have one version at a time which is not scalable for future versions of the CRD API.
* It directly depends on the client and flows a lot better by having it as a receiver of the client.
* Fits nicely as a resource verb (ensure) in that API.

If tradition is the only reason why we're hesitant about it, I'd question what the definition would be for a method that would be a good fit for the expansion API.

Concerning the slice of CRDs, I guess I don't object to just operating on a single one and letting the client orchestrate multiple. They'd only have to add a few lines of code anyways.

sttts replied:
> Can be versioned with the rest of the API. By keeping it separate in a utils pkg, you can only have one version at a time which is not scalable for future versions of the CRD API.

It can still be a helper inside of the package. Just not a method on the client.

> It directly depends on the client and flows a lot better by having it as a receiver of the client.

Not a big difference to call `client.Ensure` or `clienthelper.Ensure(client, ...)`.

> Fits nicely as a resource verb (ensure) in that API.

I don't think so. Client methods correspond to API verbs and subresources. Ensure is neither an API verb, not a subresource. The expansion interface is meant to allow custom verb implementations, e.g. like podclient.Logs or podclient.Exec.

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/utils/utils.go:
```diff
@@ -0,0 +1,105 @@
+) error {
+	funcs := make([]func() error, len(crds))
+	for index, crd := range crds {
+		crd := crd.DeepCopy() // We want to process every crd, not just the last one.
```
why a deepcopy?

frankgreco commented:
@sttts #59621 must be merged before these tests will pass as the new function `ExponentialBackoffUntil` introduced in that PR is used.

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/utils/utils.go:
```diff
@@ -0,0 +1,91 @@
+// EnsureCRDEstablishment is a utility function that will ensure that a CustomResourceDefinition
+// is not only created, but ready for use by a Kubernetes cluster. If the CustomResourceDefinition
+// cannot be established within a reasonable amount of retries, an error will be returned.
+func EnsureCRDEstablishment(
```
nit: EnsureCRDEstablished

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/utils/utils.go:
```diff
@@ -0,0 +1,91 @@
+/*
```
rename the file to ensure.go

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/utils/utils.go:
```diff
@@ -0,0 +1,91 @@
+) error {
+	// Attempt to create the crd (even if it's already present).
+	// If it is present, we still need to check its status.
+	if _, err := i.CustomResourceDefinitions().Create(crd); err != nil && !apierrors.IsAlreadyExists(err) {
```
So if the given crd is different, we will ignore that? Don't like it. Clearly, we can also not easily change the existing one. I tend to move the whole creation out of this func and just call it WaitForCRDEstablished.

frankgreco replied:
> So if the given crd is different, we will ignore that?

What do you mean? There are two possible cases here (CRD exists or not). If it does exist, than because of the `!apierrors.IsAlreadyExists(err)` check, the code will fall through to the same place as it would if it didn't exist and was created.

sttts replied:
So it does not really implement "ensure" because the existing crd might be outdated.

frankgreco replied:
by _outdated_ do you mean that the CRD is in a state where it would never be established? If so, than later login in this function would catch that case and return the appropriate error.

sttts replied:
No, I mean the case where a CRD already exists with that name, but the `crd` variable has a newer version.

sttts replied:
i.e. "ensure" sounds very much like "kubectl apply" + established.

frankgreco replied:
But the version field immutable. Hence, an error would occur when trying to create v2 if v1 already existed. For example, if you `kubectl apply` a v1 crd and try to apply a v2, you'll get the error:
```sh
The CustomResourceDefinition "foos.bar.io" is invalid: spec.version: Invalid value: "v2": field is immutable
```

frankgreco replied:
Hence, to bring back into context your question:
> So if the given crd is different, we will ignore that?

No, we would not ignore it. However, an error would be returned as is consistent with `kubectl apply` as the version field is immutable.

frankgreco replied:
I am using `Create` however and so I need to check if that will overwrite the existing one. I may need to check for existence and if already present, use `Patch`

sttts replied:
I think the easiest solution is just to turn the func into a simple WaitForEstablished helper. We cannot implement an apply-like logic easily.

frankgreco replied:
Are there apis from `kubectl` that we could leverage or would it still not be the easy.

sttts replied:
It won't be easy. Event the kubectl apply mechanism is known to be quirky. I think it's much better to leave that part to the implementor. He/she knows what to change. CRDs are too important to apply some half-baked heuristic to updates.

frankgreco replied:
Alright I'm thoroughly convinced 👍 Changes made.

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/utils/ensure.go:
```diff
@@ -0,0 +1,84 @@
+
+var (
+	// retries is the maximum number of attempts that a CRD will be given to establish
+	retries = 5
```
as this is a generic package please qualify these constants. Or even inline them.

sttts commented on staging/src/k8s.io/apiextensions-apiserver/pkg/utils/ensure.go:
```diff
@@ -0,0 +1,84 @@
+// WaitForEstablished  is a utility function that will ensure that an already created
+// CustomResourceDefinition is ready for use by a Kubernetes cluster. If the CustomResourceDefinition
+// cannot be established within a reasonable amount of retries, an error will be returned.
+func WaitForEstablished(
```
can we wire that into all the tests of CRDs?

frankgreco commented:
/remove-lifecycle stale

I will finish this this week.
